File: day7/day7.ts
import fs from 'fs'

const data = fs.readFileSync('day7-example.txt', 'utf8').trim().split(',').map(s => parseInt(s, 10))

const range = (start: number, stop: number, step = 1): number[] => {
  if (start > stop) step = -1
  return Array(Math.ceil((stop - start + step) / step)).fill(start).map((x, y) => x + y * step)
}

const computeMedian = (arr: number[]) => {
  const mid = Math.floor(arr.length / 2)
  const nums = arr.sort((a, b) => a - b);
  return arr.length % 2 !== 0 ? nums[mid] : (nums[mid - 1] + nums[mid]) / 2;
};

const computeAverage = (arr: number[]) => Math.floor(arr.reduce((a, b) => a + b) / arr.length)

const triangleNumber = (num: number) => range(1, num).reduce((a, b) => a + b)

// Works but its sloooooow 🐢
const bruteForce = (arr: number[]): number => {
  let smallestCost = Infinity
  const min = Math.min(...data)
  const max = Math.max(...data)

  for (const i of range(min, max)) {
    const total = arr.reduce((count, num) => count + triangleNumber(Math.abs(num - i)), 0)
    smallestCost = Math.min(total, smallestCost)
  }

  return smallestCost
}

const useMedian = (arr: number[]) => {
  const median = computeMedian(arr)
  return arr.reduce((count, num) => count + Math.abs(num - median), 0)
}

const useAverage = (arr: number[]) => {
  const average = computeAverage(arr)
  return arr.reduce((count: number, num: number) => count + triangleNumber(Math.abs(num - average)), 0)
}

console.log("Part 1: " + useMedian(data))

console.log("Part 2: " + useAverage(data))

File: day7/day7.py
from statistics import mean, median
from functools import reduce

f = open("./day7-example.txt", "r")
data = [int(num) for num in f.readlines()[0].strip().split(",")]


# Gets nth triangle number eg. 5th number = 5 + 4 + 3 + 2 + 1
# https://en.wikipedia.org/wiki/Triangular_number
def nth_triangular_number(num):
    return reduce(lambda a, b: a + b, range(1, num + 1), 0)


def part1(arr):
    return reduce(lambda count, num: count + abs(num - median(arr)), arr, 0)


# This works for the example but for my dataset I have to swap floor for round
# I blame advent of code :)))
def part2(arr):
    avg = round(mean(arr))

    return reduce(
        lambda count, num: count + nth_triangular_number(abs(num - avg)),
        arr,
        0,
    )


if __name__ == "__main__":
    print("Part 1: " + str(part1(data)))
    print("Part 2: " + str(part2(data)))
